tensorflow_probability/python/experimental/mcmc/sample_discarding_kernel.py

```diff
@@ -0,0 +1,177 @@
+                            'inner_results'])):
+  __slots__ = ()
+
+
+class SampleDiscardingKernel(kernel_base.TransitionKernel):
+  """Appropriately discards samples to conduct thinning and burn-in.
+
+  `SampleDiscardingKernel` is a composable `TransitionKernel` that
+  applies thinning and burn-in to samples returned by its
+  `inner_kernel`. All Transition Kernels wrapping it will only
+  see non-discarded samples.
+
+  The burn-in step conducts both burn-in and one step of thinning.
+  In other words, the first call to `one_step` will skip
+  `num_burnin_steps + num_steps_between_results` samples. All
+  subsequent calls skip only `num_steps_between_results` samples.
+  """
+
+  def __init__(
+      self,
+      inner_kernel,
+      num_burnin_steps=0,
+      num_steps_between_results=0,
+      name=None):
+    """Instantiates this object.
+
+    Args:
+      inner_kernel: `TransitionKernel` whose `one_step` will generate
+        MCMC sample(s).
+      num_burnin_steps: Integer or scalar `Tensor` representing the number
+        of chain steps to take before starting to collect results.
+        Defaults to 0 (i.e., no burn-in).
+      num_steps_between_results: Integer or scalar `Tensor` representing
+        the number of chain steps between collecting a result. Only one out
+        of every `num_steps_between_samples + 1` steps is included in the
+        returned results. Defaults to 0 (i.e., no thinning).
+      name: Python `str` name prefixed to Ops created by this function.
+        Default value: `None` (i.e., "sample_discarding_kernel").
+    """
+    if tf.get_static_value(num_burnin_steps):
+      num_burnin_steps = tf.get_static_value(num_burnin_steps)
+    if tf.get_static_value(num_steps_between_results):
+      num_steps_between_results = tf.get_static_value(
+          num_steps_between_results)
+    self._parameters = dict(
+        inner_kernel=inner_kernel,
+        num_burnin_steps=num_burnin_steps,
+        num_steps_between_results=num_steps_between_results,
+        name=name or 'sample_discarding_kernel'
+    )
+
+  def _num_samples_to_skip(self, call_counter):
+    """Calculates how many samples to skip based on the call number."""
+    # If `self.num_burnin_steps` is statically known to be 0,
+    # `self.num_steps_between_results` will be returned outright.
+    if self.num_burnin_steps == 0:
+      return self.num_steps_between_results
+    else:
+      return (tf.where(tf.equal(call_counter, 0), self.num_burnin_steps, 0) +
+              self.num_steps_between_results)
+
+  def one_step(self, current_state, previous_kernel_results, seed=None):
+    """Collects one non-discarded sample.
+
+    Args:
+      current_state: `Tensor` or Python `list` of `Tensor`s
+        representing the current state(s) of the Markov chain(s),
+      previous_kernel_results: `SampleDiscardingKernelResults` named tuple.
+        `SampleDiscardingKernelResults` contains the `call_counter`
```

Comment:
It's not necessary to inline the definition of `SampleDiscardingKernelResults`, it only increases maintenance burden.